File: mobile/lib/protocol/input_snapshot_payload.dart
import 'dart:typed_data';

import 'frame.dart';
import 'input_event.dart';
import 'payload_codec.dart';

class InputSnapshotPayload {
  const InputSnapshotPayload({required this.events});

  final List<InputEvent> events;

  @override
  bool operator ==(Object other) =>
      other is InputSnapshotPayload &&
      events.length == other.events.length &&
      _eventsEqual(events, other.events);

  @override
  int get hashCode => Object.hashAll(events);
}

class InputSnapshotPayloadCodec {
  static const int minEntries = 1;
  static const int maxEntries = 1024;

  static Uint8List encode(InputSnapshotPayload payload) {
    if (payload.events.length < minEntries || payload.events.length > maxEntries) {
      throw const ProtocolException(
        'INPUT_SNAPSHOT must contain 1..1024 entries.',
      );
    }
    final writer = PayloadWriter()..writeUInt16(payload.events.length);
    for (final event in payload.events) {
      InputEventCodec.encodeTo(event, writer);
    }
    return writer.toBytes();
  }

  static InputSnapshotPayload decode(List<int> bytes) {
    final reader = PayloadReader(bytes);
    final count = reader.readUInt16();
    if (count < minEntries || count > maxEntries) {
      throw const ProtocolException(
        'INPUT_SNAPSHOT must contain 1..1024 entries.',
      );
    }
    final events = <InputEvent>[];
    for (var i = 0; i < count; i++) {
      events.add(InputEventCodec.decodeFrom(reader));
    }
    reader.expectEnd();
    return InputSnapshotPayload(events: events);
  }
}

bool _eventsEqual(List<InputEvent> a, List<InputEvent> b) {
  for (var i = 0; i < a.length; i++) {
    if (a[i] != b[i]) {
      return false;
    }
  }
  return true;
}

File: desktop/Protocol/InputSnapshotPayload.cs
namespace CTRL.Desktop.Protocol;

public sealed record InputSnapshotPayload(IReadOnlyList<InputEvent> Events);

public static class InputSnapshotPayloadCodec
{
    public const ushort MinEntries = 1;
    public const ushort MaxEntries = 1024;

    public static byte[] Encode(InputSnapshotPayload payload)
    {
        if (payload.Events.Count < MinEntries || payload.Events.Count > MaxEntries)
            throw new ProtocolException("INPUT_SNAPSHOT must contain 1..1024 entries.");
        var writer = new PayloadWriter();
        writer.WriteUInt16((ushort)payload.Events.Count);
        foreach (var e in payload.Events)
            InputEventCodec.EncodeTo(e, writer);
        return writer.ToArray();
    }

    public static InputSnapshotPayload Decode(byte[] payload)
    {
        var reader = new PayloadReader(payload);
        var count = reader.ReadUInt16();
        if (count < MinEntries || count > MaxEntries)
            throw new ProtocolException("INPUT_SNAPSHOT must contain 1..1024 entries.");
        var events = new List<InputEvent>(count);
        for (var i = 0; i < count; i++)
            events.Add(InputEventCodec.DecodeFrom(reader));
        reader.ExpectEnd();
        return new InputSnapshotPayload(events);
    }
}
